### PR TITLE
[WIP] Add mechanism to intercept requests

### PR DESCRIPTION
**Pull request by ishkawa**
This PR implements a mechanism to intercept each requests for following purposes:

- Insert common operations before/after each requests.
- Manage network indicator visibility.

An interceptor type looks like below:

```swift
class MyRequestInterceptor: RequestInterceptorType {
    func interceptBeforeRequest<T: RequestType>(request: T) {
        /* To be called before each requests */
    }
    
    func interceptAfterRequest<T: RequestType>(request: T, result: Result<T.Response, APIError>) {
        /* To be called after each requests */
    }
}
```

Then, append its instance to `requestInterceptors` of an instance of `Session`:

```swift
let session = /* Get the session to add interceptors. */
session.requestInterceptors.append(interceptor)
```

### CODE REVIEW

**Comment by ishkawa** [comment]
Ready for review :sparkles:

**Comment by ikesyo** [comment]
My thought is that this is not a _real interceptor_, but just a callback or an observer, since this can not modify the given request nor prepend / append additional requests to the original request. [OkHttp's `Interceptor`](https://github.com/square/okhttp/wiki/Interceptors) is in my head.

IMHO we should reconsider the functionality or its naming.

**Comment by ikesyo** [comment]
> since this can not modify the given request

This is not the case if the concrete request type has settable `var` properties. :bow:

**Comment by ishkawa** [comment]
Unlike Java, requests may be value type, so I think modifying requests here is little bit complicated. IMO, common configurations of requests should be described in request protocols.

> IMHO we should reconsider the functionality or its naming.

For the reasons above, I think just changing name is better for us. `RequestObserverType` is the most likely candidate in my head. What do you think?

**Comment by ikesyo** [comment]
> For the reasons above, I think just changing name is better for us. `RequestObserverType` is the most likely candidate in my head. What do you think?

That seems reasonable! :+1:

**Comment by ishkawa** [comment]
Done!

File: APIKit/RequestObserverType.swift [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import Foundation
```
Comment (ikesyo) [inline APIKit/RequestObserverType.swift]: This import seems not to be required.

File: APIKitTests/RequestTypeTests.swift [hunk 2]
```diff
@@ -95,6 +96,42 @@ class RequestTypeTests: XCTestCase {
         waitForExpectationsWithTimeout(1.0, handler: nil)
     }
     
+    func testInterceptRequest() {
```
Comment (ikesyo) [inline APIKitTests/RequestTypeTests.swift]: `testObserveRequest`?

File: APIKit/RequestObserverType.swift [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+import Result
+
+public protocol RequestObserverType {
```
Comment (ikesyo) [inline APIKit/RequestObserverType.swift]: How about adding empty default implementations? That enables users to implement (override) only one handler they need.

File: APIKit/Session.swift [hunk 4]
```diff
@@ -3,16 +3,32 @@ import Result
...
+        self.requestObservers = requestObservers
     }
 
     // send request and build response object
     public func sendRequest<T: RequestType>(request: T, handler: (Result<T.Response, APIError>) -> Void = {r in}) -> NSURLSessionDataTask? {
+        func callbackHandler(result: Result<T.Response, APIError>) {
+            dispatch_async(dispatch_get_main_queue()) {
+                handler(result)
+                
+                for requestObserver in self.requestObservers {
```
Comment (ikesyo) [inline APIKit/Session.swift]: `forEach` can be used here (but I'm not sure which is preferable).

File: APIKit/Session.swift [hunk 5]
```diff
@@ -3,16 +3,32 @@ import Result
...
+                    requestObserver.handleAfterRequest(request, result: result)
+                }
+            }
+        }
+        
+        for requestObserver in requestObservers {
```
Comment (ikesyo) [inline APIKit/Session.swift]: As above.

**Comment by ikesyo** [comment]
LGTM other than the nitpicks. :star:

File: APIKit/RequestObserverType.swift (hunk 3 above)
Comment (ishkawa) [inline APIKit/RequestObserverType.swift]: Sounds nice :bulb:

**Comment by ishkawa** [comment]
Fixed minor issues pointed out! Thanks :+1:

**Comment by ishkawa** [comment]
Pend this PR temporarily :bow:  I found more considerations about intercepting each requests.

**Comment by ishkawa** [comment]
My thoughts have been changed to modify requests and responses. I am trying to implement *real* interceptor like below:

```swift
public protocol RequestInterceptorType {
    func interceptURLRequest<T: RequestType>(URLRequest: NSMutableURLRequest, request: T) -> NSMutableURLRequest
    func interceptResponseData<T: RequestType>(data: NSData, URLResponse: NSHTTPURLResponse, request: T) -> Result<NSData, APIError>
    func interceptResponseObject<T: RequestType>(object: AnyObject, URLResponse: NSHTTPURLResponse, request: T) -> Result<AnyObject, APIError>
}
```

This interceptor supports following usages:

- Set common configurations to each requests.
- Post NSNotification if parsed object contains `unread` key. 
- Change results depend on NSData, parsed AnyObject or NSHTTPURLResponse.  
  (e.g. always fail if the value for `X-Success` custom HTTP header is `false`.)

The first usage is also supported by request protocol, but the rest usage can't be expressed in request protocol. Additionally, there is still room for consideration of naming, I think `SessionTaskInterceptorType` would be better.

@ikesyo Do you have any thoughts on this?

**Comment by ikesyo** [comment]
@ishkawa I'm sorry for the late response! :bow: 

> - Change results depend on NSData, parsed AnyObject or NSHTTPURLResponse.
(e.g. always fail if the value for X-Success custom HTTP header is false.)

That is a good point and I really like the idea! :+1: This would be also useful for some web APIs which have a status flag in their JSON response (e.g. [Slack](https://api.slack.com/web)).

**Comment by ishkawa** [comment]
Closing this in favor of #132 and #127.

- Hooking all requests and responses is supported in #132.
- Changing `Result<Request.Response, APIError>` depending on `AnyObject` and `NSHTTPURLResponse` is supported by `validateObject(_:URLResponse:)` in #127.

**Pull request closed without merging**